File: python/tests/test_backbone.py
from __future__ import annotations

import sys
import unittest
from pathlib import Path

SRC = Path(__file__).resolve().parents[1] / "src"
sys.path.insert(0, str(SRC))

from kinetic_mhd_pinn import CoupledState, MODULE_SPECS, nominal_startup_order
from kinetic_mhd_pinn.common.registry import validate_registry
from kinetic_mhd_pinn.common.scaffold import ScaffoldModule


class BackboneTests(unittest.TestCase):
    def test_registry_is_valid(self) -> None:
        validate_registry()
        self.assertEqual(len(MODULE_SPECS), 11)

    def test_nominal_order_is_numbered(self) -> None:
        self.assertEqual(nominal_startup_order(), tuple(f"module_{i}" for i in range(1, 12)))

    def test_shared_state_contract(self) -> None:
        state = CoupledState()
        module = ScaffoldModule(module_id="module_test", produced_marker="test_product")
        module.run(state)
        self.assertTrue(state.fields["test_product"])
        self.assertEqual(state.history[-1]["module"], "module_test")

    def test_missing_product_fails_loudly(self) -> None:
        state = CoupledState()
        module = ScaffoldModule(module_id="module_test", required_products=("missing",))
        with self.assertRaises(KeyError):
            module.run(state)


if __name__ == "__main__":
    unittest.main()

File: matlab/tests/test_backbone.m
function tests = test_backbone
%TEST_BACKBONE Contract tests for the MATLAB project scaffold.
tests = functiontests(localfunctions);
end

function setupOnce(testCase)
testsDir = fileparts(mfilename('fullpath'));
matlabRoot = fileparts(testsDir);
sourceDir = fullfile(matlabRoot, 'src');
addpath(sourceDir);
testCase.TestData.sourceDir = sourceDir;
end

function teardownOnce(testCase)
rmpath(testCase.TestData.sourceDir);
end

function testRegistryIsValid(testCase)
kinetic_mhd_pinn.common.validate_registry();
specs = kinetic_mhd_pinn.common.registry();
verifyEqual(testCase, numel(specs), 11);
end

function testNominalOrderIsNumbered(testCase)
actual = kinetic_mhd_pinn.common.nominal_startup_order();
expected = arrayfun(@(i) sprintf('module_%d', i), 1:11, ...
    'UniformOutput', false);
verifyEqual(testCase, actual, expected);
end

function testSharedStateContract(testCase)
state = kinetic_mhd_pinn.common.CoupledState();
module = kinetic_mhd_pinn.common.ScaffoldModule( ...
    'module_test', {}, 'test_product');
module.run(state);
verifyTrue(testCase, state.fields.test_product);
verifyEqual(testCase, state.history{end}.module, 'module_test');
end

function testMissingProductFailsLoudly(testCase)
state = kinetic_mhd_pinn.common.CoupledState();
module = kinetic_mhd_pinn.common.ScaffoldModule( ...
    'module_test', {'missing'}, 'unused');
verifyError(testCase, @() module.run(state), ...
    'kinetic_mhd_pinn:MissingProducts');
end

function testAllModuleScaffoldsAreAddressable(testCase)
specs = kinetic_mhd_pinn.common.registry();
for i = 1:numel(specs)
    factory = str2func(sprintf('kinetic_mhd_pinn.%s.solver', specs(i).package));
    module = factory();
    verifyEqual(testCase, module.moduleId, specs(i).id);
end
end
